### ISSUE TITLE
Allow aborting/skipping one pam module

### ISSUE THREAD

**Issue by nomeata** [google/xsecurelock#145]
I started using the fingerprint scanner on my laptop, with a pam configuration like this:
```
auth sufficient /nix/store/ak1j194wmjb5py2820x4a7rlmiv2zz4f-fprintd-1.94.2/lib/security/pam_fprintd.so
auth sufficient pam_unix.so nullok  likeauth try_first_pass
auth required pam_deny.so
```

With, say, `sudo` I can use Ctrl-C to abort the fingerprint prompt if I want to use my password (say, my finger is bandaged, or the fingerprint scanner is somehow unreachable).

With `xsecurelock`, the best I can do is to use a wrong finger three times to make the first pam entry fail, and then I can use my password.

It would be nice if I could instruct `xsecurelock` to abort the fingerprint reader prompt, e.g. by pressing Esc.

**Comment by divVerent** [comment]
This is an interesting request; I do not yet know how this can work though, as PAM does not have a cancel API. Maybe pam_fprintd just catches SIGINT?

I will check out source code of pam_fprintd to find out more about this.

**Comment by nomeata** [comment]
Yes, I think that's precisely how these pam modules work around that issue. But better check, I am only relaying heresay :-)

**Comment by divVerent** [comment]
Can't find anything on https://github.com/dsd/fprintd/tree/master/pam that handles SIGINT. Too bad I don't have a FP reader myself, or I could try with that.

**Comment by divVerent** [comment]
OK, so we know it is not that simple.

I do wonder how the Ctrl+C then works. Can you reproduce it with sudo as
follows:

- sudo something
wait for fingerprint prompt

now do NOT press the finger, but run "ps waxwuf | grep -10 sudo"

Does that show any subprocesses related to pam_fprintd?

**Comment by nomeata** [comment]
In https://gitlab.freedesktop.org/libfprint/fprintd/-/tags/v1.92.0 I see 
> * pam: Cancel authentication on SIGINT (e.g. ctrl+c with sudo)

this is the commit I think https://github.com/freedesktop/libfprint-fprintd/commit/657f58fd648e35417ce7266b9c1558ce497dc179

**Comment by divVerent** [comment]
But then the "pkill -INT authproto_pam" way should have worked...

That is odd then.

**Comment by nomeata** [comment]
Maybe the problem isn’t that the pam module doesn't handle it, but that `authroto_pam` _also_ gets the signal, and dies, while `sudo` (just guessing here) doesn’t die with SIGINT while pam is running.

**Comment by nomeata** [comment]
Indeed, I can make `sudo` behave that way not just with Ctrl-C, but also with `sudo kill -INT $(pgrep sudo)` (I need to run this as root because `sudo` runs as root)

**Comment by nomeata** [comment]
https://www.sudo.ws/repos/sudo/rev/524d95ac222e might be relevant

**Comment by divVerent** [comment]
Can you try the change I just pushed to branch ctrl-c?

**Comment by nomeata** [comment]
Hmm, doesn't quite seem to work.

I start `xsecurelock`, move the mouse to be prompted for the finger print. Then I press Ctrl-C. 
At this point, nothing changes in the view, but the finger print doesn’t do anything (my hypothesis: the auth helper is killed, but the display is not redrawn).
If I now move the mouse again, the prompt appears again, slightly moved (a feature of xsecurelock), and I am asked to put the fingerprint there.

When I unlock, the console says 
```
2022-10-11T09:17:19Z 845286 xsecurelock: auth child killed by signal 2.
```

So it looks like Ctrl-C takes the auth child down completely, rather than just interrupting one pam interaction?

**Comment by divVerent** [comment]
That is helpful - gotta investigate further.
